---
title: "QC and creating seurat object with all metadata, Exp1 aligned to mm10-liftoff"
author: "Hope Lab snRNA team"
date: "07/07/2023"
output:
  word_document: default
  fig_width: 6
  fig_height: 4
---
## Set knitting settings

```{r Setup, include=FALSE, echo = FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Load packages

```{r LoadPackages}
library(Seurat)
library(ggplot2)
```

## Set working directory. Import data, create Seurat object, and assign sample tag call to each nuclei.

```{r import}
# set working directory
setwd("~/Documents/GitHub/XPoSE/Scripts/")
# Read gene count data tables in 

counts.C1 <- read.table("~/Desktop/Analysis/XPoSE_seq_cartridge/Combined_cart1-good-only_RSEC_MolsPerCell.csv",
                        sep = ",", skip = 7, header = TRUE, row.names = 1)
counts.C2 <- read.table("~/Desktop/Analysis/XPoSE_seq_cartridge/Combined_cart2-good-only_RSEC_MolsPerCell.csv",
                        sep = ",", skip = 7, header = TRUE, row.names = 1)

# store count tables as a list
counts <- list(c1 = counts.C1, c2 = counts.C2)

# Import and add sample tag identification and format to one column
tags.C1 <- read.table("~/Desktop/Analysis/XPoSE_seq_cartridge/cart1-good-only_Sample_Tag_Calls.csv", sep = ",", skip = 7, header = TRUE, row.names = 1)
tags.C2 <- read.table("~/Desktop/Analysis/XPoSE_seq_cartridge/cart2-good-only_Sample_Tag_Calls.csv", sep = ",", skip = 7, header = TRUE, row.names = 1)

tags <- list(t1 = subset(tags.C1, select=-c(2)),
             t2 = subset(tags.C2, select=-c(2)))

source("Functions/create_seur.R")

# creates separate seurat object for each count table present in environment
for(i in 1:length(counts)){
   assign(paste("data",".C",i,sep=""),
          create_seur(counts[i], i, tags[i]),
          envir = .GlobalEnv)
   
}

# Creating a list of both Seurat Objects for easier reference
seur_list <- list(data.C1 = data.C1, data.C2 = data.C2) # this needs to be changed if we have more than 1 cartridge

```


## Cartridge QC plots, looking at number of genes and transcripts detected

```{r QCplots}

source("Functions/plot_detection.R")

for(i in 1:length(seur_list)){
  for(f in c("genes", "transcripts")){
    print(plot_detection(f, i))
  }
}

```

## Break each sample tag into an individual object and assign meta data. Combine back together as single object

```{r AssignAndCombine}

# input list of ratIDs used
ratIDs <- c(NA, "KES029R7", "KES029R1", "KES029R6", "KES029R5", "KES029R8",
            "KES029R2", "KES016R12", "KES016R10")

# input sex of each rat
rat_sex <- c(NA, "male", "male", "female", "female", "male",
             "male", "female", "female")

# creates sample tag list
st_list <- c(NA)
for(i in 2:length(ratIDs)){
  st_list <- c(st_list,paste("SampleTag0",i,"_mm",sep=""))
}
  
source("Functions/assign_combine.R") # read function for assignment

for (l in 1:length(seur_list)) {
  seur_list <- assign_combine(seur_list, l, ratIDs, st_list, rat_sex)
}
```
